### PR TITLE
Add decimal TRUNCATE(``x``, ``n``) function.

### PR DESCRIPTION
**Pull request by fiedukow**
This function takes `x` value DECIMAL(p, s) and
return the value of same type truncated to `n` places
after decimal point. All decimal places starting at `n`
after decimal point are zeroed.

In case of negative `n` all decimal places after decimal
point as well as abs(`n`) decimal places before decimal
point are truncated.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 1]
```diff
@@ -526,6 +527,69 @@ private static SqlScalarFunction decimalRoundNFunction()
                 .build();
     }
 
+    private static SqlScalarFunction decimalTruncateFunction()
```
Comment (ilfrin) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: (1) don't insert methods for a function in between methods that compose a different function (roundN). 
(2) either use xxTruncateFunction or xxTruncateNFunction, not both if there is only one function implementation

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 2]
```diff
@@ -526,6 +527,69 @@ private static SqlScalarFunction decimalRoundNFunction()
...
+    {
+        Signature signature = Signature.builder()
+                .kind(SCALAR)
+                .name("truncate")
+                .literalParameters("num_precision", "num_scale", "result_precision")
+                .longVariableConstraints(
+                        longVariableCalculation("result_precision", "min(38, num_precision)"))
+                .argumentTypes("decimal(num_precision, num_scale)", BIGINT)
+                .returnType("decimal(result_precision, num_scale)")
```
Comment (ilfrin) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: can the precision be bigger than original? 

can't we do a better job with the result scale?

Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: Precision should never increase.
We can reduce output scale to be `Max(0, n)` and precision to be `num_prec - Max(0, n)`, but I thought it's not worth changing the number type in "rounding" function. Do you wan't me to do that?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 1 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: Both fixed.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java [hunk 3]
```diff
@@ -545,6 +545,46 @@ public void testRound()
...
+        assertFunction("truncate(DECIMAL '1234', -1)", createDecimalType(4, 0), SqlDecimal.of("1230"));
+        assertFunction("truncate(DECIMAL '1234.56', 1)", createDecimalType(6, 2), SqlDecimal.of("1234.50"));
+        assertFunction("truncate(DECIMAL '1234.56', -1)", createDecimalType(6, 2), SqlDecimal.of("1230.00"));
+        assertFunction("truncate(DECIMAL '-1234.56', 1)", createDecimalType(6, 2), SqlDecimal.of("-1234.50"));
+
+        assertFunction("truncate(DECIMAL '1234', -4)", createDecimalType(4, 0), SqlDecimal.of("0000"));
+        assertFunction("truncate(DECIMAL '1234.56', -4)", createDecimalType(6, 2), SqlDecimal.of("0000.00"));
+        assertFunction("truncate(DECIMAL '-1234.56', -4)", createDecimalType(6, 2), SqlDecimal.of("0000.00"));
+
+        assertFunction("truncate(DECIMAL '1234', 1)", createDecimalType(4, 0), SqlDecimal.of("1234"));
```
Comment (ilfrin) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java]: this line is the same as the one at the top

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java [hunk 4]
```diff
@@ -545,6 +545,46 @@ public void testRound()
...
+        assertFunction("truncate(DECIMAL '123456789012345678901', -1)", createDecimalType(21, 0), SqlDecimal.of("123456789012345678900"));
+        assertFunction("truncate(DECIMAL '123456789012345678901.23', 1)", createDecimalType(23, 2), SqlDecimal.of("123456789012345678901.20"));
+        assertFunction("truncate(DECIMAL '123456789012345678901.23', -1)", createDecimalType(23, 2), SqlDecimal.of("123456789012345678900.00"));
+        assertFunction("truncate(DECIMAL '123456789012345678999.99', -1)", createDecimalType(23, 2), SqlDecimal.of("123456789012345678990.00"));
+        assertFunction("truncate(DECIMAL '-123456789012345678999.99', -1)", createDecimalType(23, 2), SqlDecimal.of("-123456789012345678990.00"));
+
+        assertFunction("truncate(DECIMAL '123456789012345678901', -21)", createDecimalType(21, 0), SqlDecimal.of("000000000000000000000"));
+        assertFunction("truncate(DECIMAL '123456789012345678901.23', -21)", createDecimalType(23, 2), SqlDecimal.of("000000000000000000000.00"));
+
+        assertFunction("truncate(DECIMAL '123456789012345678901', 1)", createDecimalType(21, 0), SqlDecimal.of("123456789012345678901"));
```
Comment (ilfrin) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java]: same as the first "LongDecimal" asserts

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 2 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: I think we cannot really make such constraint using function builder, as the `n` value is unknown at that stage.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java (hunk 3 above)
Comment (fiedukow) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java]: done

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java (hunk 4 above)
Comment (fiedukow) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java]: done

**Comment by ilfrin** [comment]
overall lgtm, go though the test cases though and check if all of them are testing for something, I guess you copied them from other cases where there were boundary crossings (long-short), we don't have it here so those tests would not be necessary

**Comment by fiedukow** [comment]
Rebased version available in #163.

**Pull request closed without merging**